client/src/lib/src/availableVariables.ts

```diff
@@ -0,0 +1,369 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+// Variables made available from /poky/plain/meta/conf/bitbake.conf
+export const commonDirectoriesVariables = new Set([
+  'base_prefix',
+  'prefix',
+  'exec_prefix',
+  'base_bindir',
+  'base_sbindir',
+  'base_libdir',
+  'datadir',
+  'sysconfdir',
+  'servicedir',
+  'sharedstatedir',
+  'localstatedir',
+  'infodir',
+  'mandir',
+  'docdir',
+  'bindir',
+  'sbindir',
+  'libexecdir',
+  'libdir',
+  'includedir',
+  'oldincludedir',
+  'base_bindir_native',
+  'base_sbindir_native',
+  'sysconfdir_native',
+  'prefix_native',
+  'bindir_native',
+  'sbindir_native',
+  'includedir_native',
+  'libdir_native',
+  'datadir_native',
+  'bindir_cross',
+  'target_datadir',
+  'BUILD_ARCH',
```

Comment:
Yes, they are all merged together in priority: https://github.com/yoctoproject/vscode-bitbake/blob/1ec99293b6b9665df8f2e09a9d12a4fb614b1960/server/src/connectionHandlers/onCompletion.ts#L142

There's a test that ensures [there is no duplicates](https://github.com/yoctoproject/vscode-bitbake/blob/1ec99293b6b9665df8f2e09a9d12a4fb614b1960/server/src/__tests__/completions.test.ts#L899), and a test that ensures the ones Ziwei made [have priority](https://github.com/yoctoproject/vscode-bitbake/blob/1ec99293b6b9665df8f2e09a9d12a4fb614b1960/server/src/__tests__/completions.test.ts#L900) (since they have more information)

Reply:
Can't you get rid of this list then?

Reply:
Some of these variables are not available into the [Yocto](https://docs.yoctoproject.org/ref-manual/variables.html) or [Bitbake](https://docs.yoctoproject.org/bitbake/bitbake-user-manual/bitbake-user-manual-ref-variables.html) variables glossaries (`DEBIAN_MIRROR`, for example). Therefore they would not be available if we get rid of the list. 

I don't think we should put too much efforts in having this list cleaner and avoiding duplicates. By doing so we might accidentally delete variables that should not be deleted, and we would complexify the process of updating it.

I first wanted to push a script that would extract the variables from `meta/conf/bitbake.conf`, but I thought we're better to consider this PR simply as a first iteration before a more sophisticated solution. I have in mind we should use the result of `bitbake -e` instead in order to get more accurate results. That way we could ensure all the variables that are available to the user would be comprised in the completion items, and nothing more. We could even give an error when the user tries to use a variable that is not available.

Reply:
`bitbake -e` is the way to go. You'll also get final values for hover and definitions.
I don't think you should add errors if not available. At most a diagnostic warning. It is valid to use an unset variables, which translates into an empty string.